AutoCommissioner: Set default NTP to internal buffer

cecille opened the pull request:
Also, it would be good to actually set the time zone to our own buffer.

cecille committed:
AutoCommissioner: Set default NTP to internal buffer

Also, it would be good to actually set the time zone to our own
buffer.

cecille committed:
Merge branch 'master' into autocommissioner_default_ntp

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -197,13 +197,28 @@ CHIP_ERROR AutoCommissioner::SetCommissioningParameters(const CommissioningParam
             if (mTimeZoneBuf[i].name.HasValue())
             {
                 auto span = MutableCharSpan(mTimeZoneNames[i], kMaxTimeZoneNameLen);
+                // This buffer is statically allocated and if of size kMaxSupportedTimeZones, so this should never fail
```
The question is whether the buffer currently in `mTimeZoneBuf[i].name.Value()` has length greater than `kMaxTimeZoneNameLen`.  I don't think anything guarantees that.

cecille replied:
I actually changed my mind here about what to do in this case. Let me just get through this meeting run and I'm going to change this and see if I can add a test.

cecille replied:
sorry, forgot to actually send the comments on this one. I've updated this. Attempting to copy anything to the buffer and letting it be truncated is not worthwhile because neither the time zone nor the ntp address can be used if they are partial. Also, for time zone, it's definitely wrong. So I've just blocked them. Right now, that field just gets omitted, but I'm also open to erroring.

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -197,13 +197,28 @@ CHIP_ERROR AutoCommissioner::SetCommissioningParameters(const CommissioningParam
+            app::DataModel::Nullable<CharSpan> default_ntp(CharSpan(mDefaultNtp, size));
+            mParams.SetDefaultNTP(default_ntp);
+        } else {
+            mParams.SetDefaultNTP(app::DataModel::Nullable<CharSpan>());
```
```suggestion
            mParams.SetDefaultNTP(NullNullable);
```
should work, I think.

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -197,13 +197,28 @@ CHIP_ERROR AutoCommissioner::SetCommissioningParameters(const CommissioningParam
+            size_t size = std::min(params.GetDefaultNTP().Value().Value().size(), kMaxDefaultNtpSize);
+            // This buffer is statically allocated and is of size kMaxDefaultNtpSize.
+            auto span = MutableCharSpan(mDefaultNtp, kMaxDefaultNtpSize);
+            CopyCharSpanToMutableCharSpan(params.GetDefaultNTP().Value().Value(), span);
```
This copy will fail if `params.GetDefaultNTP().Value().Value().size() > kMaxDefaultNtpSize`, no?   And then we have garbage in `span`....

What we might want to do is grab `params.GetDefaultNTP().Value().Value()` into an on-stack Span, figure out the size we want to copy, take a subspan of that size, and copy that....

cecille replied:
As above - I think we need a different approach here because a partial string is basically useless. Going to do something different, and I THINK I can add a unit test.

bzbarsky-apple commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -197,13 +197,28 @@ CHIP_ERROR AutoCommissioner::SetCommissioningParameters(const CommissioningParam
+            // This buffer is statically allocated and is of size kMaxDefaultNtpSize.
+            auto span = MutableCharSpan(mDefaultNtp, kMaxDefaultNtpSize);
+            CopyCharSpanToMutableCharSpan(params.GetDefaultNTP().Value().Value(), span);
+            app::DataModel::Nullable<CharSpan> default_ntp(CharSpan(mDefaultNtp, size));
+            mParams.SetDefaultNTP(default_ntp);
```
```suggestion
            auto default_ntp = MakeNullable(CharSpan(mDefaultNtp, size));
            mParams.SetDefaultNTP(default_ntp);
```

cecille committed:
Apply suggestions from code review

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

cecille committed:
Size check and fail big strings, add tests, fix bug

- neither DefaultNTP nor the time zone name is useful to the device
  if they're truncated, so don't attempt to send anything unless the
  value is spec compliant
- NOTE: Right now, we still attempt to commission with the remaining
        parameters, but I'm open to returning an error here too
- add tests for too-long names AND max size names
- fix a bug in the server default NTP handling that prevented
  domain names from ever being set
- turn on dns resolution in the all-clusters. Note that the all
  clusters app will never actually DO anything with these names,
  but it's still nice to have it available for testing.

restyled-commits committed:
Restyled by clang-format

cecille committed:
Apply suggestions from code review

Co-authored-by: Boris Zbarsky <bzbarsky@apple.com>

cecille committed:
Address review comments

restyled-commits committed:
Restyled by clang-format

cecille commented on src/controller/AutoCommissioner.cpp:
```diff
@@ -194,16 +198,35 @@ CHIP_ERROR AutoCommissioner::SetCommissioningParameters(const CommissioningParam
             }
+            else
+            {
+                mTimeZoneBuf[i].name = NullOptional;
```
As above, might as well do it.

cecille commented on src/app/clusters/time-synchronization-server/time-synchronization-server.cpp:
```diff
@@ -1284,24 +1284,19 @@ bool emberAfTimeSynchronizationClusterSetDefaultNTPCallback(
-        if (GetDelegate()->IsNTPAddressDomain(dNtpChar.Value()))
+        bool isDomain = GetDelegate()->IsNTPAddressDomain(dNtpChar.Value());
+        bool isIPv6   = GetDelegate()->IsNTPAddressValid(dNtpChar.Value());
+        bool useable  = dnsResolve ? (isIPv6 || isDomain) : isIPv6;
```
It is. And I'm taking another trip through the CI anyway because I added the comment changes so may as well change. Addressed in next commit.

cecille committed:
Merge remote-tracking branch 'upstream/master' into autocommissioner_default_ntp

cecille committed:
Merge branch 'master' into autocommissioner_default_ntp

cecille committed:
Merge branch 'master' into autocommissioner_default_ntp
